MoreLinq/CountBy.cs

```diff
@@ -0,0 +1,115 @@
+        /// <typeparam name="TKey">Type of the projected element.</typeparam>
+        /// <param name="source">Source sequence.</param>
+        /// <param name="keySelector">Function that transforms each item of source sequence into a key to be compared against the others.</param>
+        /// <param name="comparer">The equality comparer to use to determine whether or not keys are equal.
+        /// If null, the default equality comparer for <c>TSource</c> is used.</param>
+        /// <returns>A sequence of unique keys and their number of occurrences in the original sequence.</returns>
+        public static IEnumerable<KeyValuePair<TKey, int>> CountBy<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey> comparer)
+        {
+            if (source == null) throw new ArgumentNullException("source");
+            if (keySelector == null) throw new ArgumentNullException("keySelector");
+
+            return CountByImpl(source, keySelector, comparer);
+        }
+
+        private static IEnumerable<KeyValuePair<TKey, int>> CountByImpl<TSource, TKey>(IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey> comparer)
+        {
+            var dic = new Dictionary<TKey, int>(comparer);
+            var keys = new List<TKey>();
+            var counts = new List<int>();
+            var havePrevKey = false;
+            var prevKey = default(TKey);
+            var index = 0;
+
+            foreach (var item in source)
+            {
+                var key = keySelector(item);
+
+                if (// key same as the previous? then re-use the index
+                    (havePrevKey && dic.Comparer.GetHashCode(prevKey) == dic.Comparer.GetHashCode(key)
+                                  && dic.Comparer.Equals(prevKey, key))
+                    // otherwise try & find index of the key
+                    || dic.TryGetValue(key, out index))
+                {
+                    counts[index]++;
+                }
+                else
+                {
+                    dic[key] = keys.Count;
+                    keys.Add(key);
+                    counts.Add(1);
+                }
+
+                prevKey = key;
+                havePrevKey = true;
+            }
+
+            // The dictionary is no longer needed from this point forward so
+            // lose the reference and make it available as food for the GC.
+            // This optimization is designed to help cases where a slow running
+            // loop over the yielded pairs could span GC cycles. However,
+            // instead of doing simply the following:
+            //
+            // dic = null;
+            //
+            // the reference is nulled through a method that the JIT compiler
+            // is told not to try and inline; done so assuming that the above
+            // method could have been turned into a NOP (in theory).
+
+            Null(ref dic); // dic = null;
```

Comment:
I think it would be less magic and more readable to simply split the counting loop to a separate function.

For other readers (and myself, I had to google this), setting `dic = null` is usually not useful, because the CLR knows that you are not using the variable anymore, and thus marks the referenced object as eligible for collection. However, the CLR cannot be sure when a generator or async method is used (I only found a passing reference by [Stephen Cleary](http://blog.stephencleary.com/2010/02/q-should-i-set-variables-to-null-to.html), is there a more detailed source?), and thus it waits until the method ends. I *think* this is because the generator method is compiled to a class, with all local parameters as fields on that class. If the instance of this generated class outlives the use of the fields in that class, the CLR cannot know this. However, this same analysis would suggest that it would be a bug in either the C# compiler or the JIT to turn `dict = null` into a NOP.